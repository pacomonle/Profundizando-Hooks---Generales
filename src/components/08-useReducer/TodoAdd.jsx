import React from 'react'
import useForm from '../../hooks/useForm'
import './styles.css'

const TodoAdd = ({ handleAddTodo}) => {

    // hook personalizado useForm para el input del form
const [formValue, reset,  handleInputChange] = useForm({
    inputValue: ''
})

const  {inputValue} = formValue

const handleSubmit = (e) => {
    e.preventDefault();
if( inputValue.trim().length < 1){ return }
    const newTodo = {
        id: new Date().getTime(),
        desc: inputValue,
        done: false
    }

handleAddTodo(newTodo)

reset()
}

    return (
        <>
        <h3>Agregar TODO</h3>
        <hr />
        <form onSubmit={handleSubmit}>
            <input 
            type="text" 
            className='form-control'
            placeholder='Aprender...'
            autoComplete= 'off'
            name="inputValue" 
            value= {inputValue}
            onChange={handleInputChange} 
            />
            <button 
            className='btn btn-outline-primary btn-block my-1'
            type="submit">
            Agregar
            </button>
        </form>
        </>
    )
}

export default TodoAdd
