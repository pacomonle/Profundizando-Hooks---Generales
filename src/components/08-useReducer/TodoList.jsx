import React from 'react'
import './styles.css'
import TodoListItem from './TodoListItem'

const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <>
        <ul className='list-group list-group-flush'>
        {
            todos.map((todo, index) => (
                <TodoListItem 
                key={todo.id}
                index={index}
                todo={todo} 
                handleDelete={handleDelete} 
                handleToggle={handleToggle}>
                </TodoListItem>
               
            ) )
        }

</ul>
        </>
    )
}

export default TodoList
