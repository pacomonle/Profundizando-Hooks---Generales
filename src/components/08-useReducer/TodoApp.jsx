import React, { useEffect, useReducer } from 'react'
import todoReducer from './todoReducer'
import './styles.css'

import TodoList from './TodoList'
import TodoAdd from './TodoAdd'



const initialState = []

const init = () => {
    
 return JSON.parse(localStorage.getItem('todos')) || []
            
}


const TodoApp = () => {



    
// reducer -es una funcion, initialState -estado inicial , init -es una funcion que trabaja el estado inicial
const [todos, dispatch] = useReducer(todoReducer, initialState, init)

useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
}, [todos])




// para poder pasar como prop la funcion al hija
const handleAddTodo = (newTodo) => {
    const addTodoAction = {
        type: 'add',
        payload: newTodo
    }

    dispatch(addTodoAction)
}



const handleDelete = (id) => {
    const deleteTodoAction = {
        type: 'delete',
        payload: id
    }
    dispatch(deleteTodoAction)
}


const handleToggle = (id) => {
console.log(id)
   
  dispatch({
    type: 'toggle',
    payload: id
   })
}

    return (
        <div>
            <h1>Hook useReducer</h1>
            <h3>Todo APP</h3><small>Total todos: {todos ? todos.length : 0}</small>
            <hr />
            <div class="row">
                <div className='col-7'>
                    <TodoList 
                    todos={todos} 
                    handleDelete={handleDelete} 
                    handleToggle={handleToggle}>
                    </TodoList>
                   
                </div>
                <div className='col-5'>
                  <TodoAdd
                  handleAddTodo={handleAddTodo}
                  ></TodoAdd>
                </div>
            </div>
            
        </div>
    )
}

export default TodoApp
