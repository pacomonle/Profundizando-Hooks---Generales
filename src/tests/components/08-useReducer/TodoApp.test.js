import React from 'react';
import { shallow, mount } from "enzyme";
import TodoApp  from "../../../components/08-useReducer/TodoApp"
import { demoTodos } from '../../fixtures/demoTodos';
import { act } from '@testing-library/react';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';
import TodoList from '../../../components/08-useReducer/TodoList';



describe('Pruebas en <TodoApp />', () => {

    const wrapper = shallow( <TodoApp />);
  // comprobar el local storage -> usal el jest.fn()
    Storage.prototype.setItem = jest.fn(()=> {});

    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('debe de agregar un TODO', () => {
      // para usar y simular funciones en contexto en general
        const wrapper = mount( <TodoApp /> );

        act( () => {
            wrapper.find(TodoAdd).prop('handleAddTodo')( demoTodos[0] );
            wrapper.find(TodoAdd).prop('handleAddTodo')( demoTodos[1] );
        });

        expect( wrapper.find('small').text().trim() ).toBe('Total todos: 2');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
        
    });

    test('debe de eliminar un todo', () => {
        
        wrapper.find(TodoAdd).prop('handleAddTodo')( demoTodos[0] );
  
        expect( wrapper.find('small').text().trim() ).toBe('Total todos: 1');

        wrapper.find(TodoList).prop('handleDelete')( demoTodos[0].id );

        expect( wrapper.find('small').text().trim() ).toBe('Total todos: 0');

    });
    
    
    
    
})
