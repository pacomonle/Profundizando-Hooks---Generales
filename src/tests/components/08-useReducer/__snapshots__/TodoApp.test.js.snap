// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Hook useReducer
  </h1>
  <h3>
    Todo APP
  </h3>
  <small>
    Total todos: 
    0
  </small>
  <hr />
  <div
    class="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
